### PR TITLE
fix #303756 and #293370: export to pdf/png/svg with the option to use the page color or the wallpaper

### PR DESCRIPTION
**Pull request by SKefalidis**
Resolves: https://musescore.org/en/node/303756 and https://musescore.org/en/node/293370

Added the ability to use the page color when exporting to PDF/PNG/SVG. The page color still is a preference. I could change that and make it part of each Page object (as proposed in the issue). Opinions on whether I should do that, or leave the page color as a preference are welcome :^)

Exporting the wallpaper seems a bit more problematic. The ScoreView changes the size of the displayed wallpaper depending on the paintEvent and zoom level. At the same time, changing the exported file depending on the zoom level in the program seems weird. I decided to use the size of the page which at least leads to consistent behavior when exporting.  

Page Color:
image
Screenshot from 2020-04-19 12-36-05

Page Wallpaper:
Screenshot from 2020-04-19 12-15-36
Screenshot from 2020-04-19 12-16-33

- [x] I signed [CLA](https://musescore.org/en/cla)
- [x] I made sure the code in the PR follows [the coding rules](https://musescore.org/en/handbook/developers-handbook/finding-your-way-around/musescore-coding-rules)
- [x] I made sure the code compiles on my machine
- [x] I made sure there are no unnecessary changes in the code
- [x] I made sure the title of the PR reflects the core meaning of the issue you are solving
- [x] I made sure the commit message(s) contain a description and answer the question "Why do those changes fix that particular issue?" or "Why are those changes really necessary as improvements?"
- [x] I made sure the commit message title starts with "fix #424242:" if there is a related issue
- [ ] I created the test (mtest, vtest, script test) to verify the changes I made

### CODE REVIEW

File: global/settings/types/preferencekeys.h [hunk 1]
```diff
@@ -112,6 +112,7 @@
 #define PREF_UI_CANVAS_FG_USECOLOR_IN_PALETTES              "ui/canvas/foreground/useColorInPalettes"
 #define PREF_UI_CANVAS_BG_COLOR                             "ui/canvas/background/color"
 #define PREF_UI_CANVAS_FG_COLOR                             "ui/canvas/foreground/color"
+#define PREF_UI_CANVAS_FG_EXPORT_PDF                        "ui/canvas/foreground/export/pdf"
```
Comment (Jojo-Schmitz) [inline global/settings/types/preferencekeys.h:115]: Isn't that rather "background"?

Comment (Jojo-Schmitz) [reply global/settings/types/preferencekeys.h:115]: And why restrict this to PDF and not having it for PNG, SVG, and possibly MusicXML too?

Comment (SKefalidis) [reply global/settings/types/preferencekeys.h:115]: I don't think that it is "background". The background color is the one behind the page.

**Comment by SKefalidis** [comment]
Added the option for svg and png.

File: global/settings/types/preferencekeys.h (hunk 1 above)
Comment (SKefalidis) [reply global/settings/types/preferencekeys.h:115]: As far as MusicXML is concerned, I will probably make some decently sized changes in the MusicXML code, so I'll add png and svg for now.

**Comment by SKefalidis** [comment]
Fixed formatting.

File: global/settings/types/preferencekeys.h (hunk 1 above)
Comment (Jojo-Schmitz) [reply global/settings/types/preferencekeys.h:115]: forgroung color is that of notes, staff lines etc, (black by default), bachound is white by default and your PR allows to change that. And not just for PDF anymore now, it seems

Comment (SKefalidis) [reply global/settings/types/preferencekeys.h:115]: image

The background color is dark-blue by default and the paper color is white. 

image

The paper color is called fgColor in prefsdialog.*

You are probably right, but I seem unable to understand my mistake. I am essentially copying the implementation of use the same color in palettes (I should change the name of my macro).

**Comment by Jojo-Schmitz** [comment]
The vtests are not happy. Some ore screwed up pretty badly

**Comment by Jojo-Schmitz** [comment]
Hmm, seems the default paper color is #f9f9f9 and not white?

**Comment by SKefalidis** [comment]
I might have caused it by changing the way the pages are painted in svg (for uniformity). I reverted those changes and I am waiting for the new vtest.

File: mscore/preferences.cpp [hunk 2]
```diff
@@ -158,6 +158,7 @@ void Preferences::init(bool storeInMemoryOnly)
             {PREF_UI_CANVAS_FG_USECOLOR_IN_PALETTES,               new BoolPreference(false, false)},
             {PREF_UI_CANVAS_BG_COLOR,                              new ColorPreference(QColor("#142433"), false)},
             {PREF_UI_CANVAS_FG_COLOR,                              new ColorPreference(QColor("#f9f9f9"), false)},
```
Comment (Jojo-Schmitz) [inline mscore/preferences.cpp:160]: I guess we'd need to interpret this color as white?

File: mscore/file.cpp [hunk 3]
```diff
@@ -2225,6 +2234,13 @@ bool MuseScore::savePdf(QList<Score*> cs_, const QString& saveName)
                   if (!firstPage)
                         printer.newPage();
                   firstPage = false;
+
+                  if(usePageColor) {
```
Comment (Jojo-Schmitz) [inline mscore/file.cpp:2238]: space missing

File: mscore/file.cpp [hunk 4]
```diff
@@ -2143,10 +2143,18 @@ bool MuseScore::savePdf(Score* cs_, QPrinter& printer)
       const QList<Page*> pl = cs_->pages();
       int pages = pl.size();
       bool firstPage = true;
+      bool usePageColor = preferences.getBool(PREF_UI_CANVAS_FG_USECOLOR_IN_EXPORT);
       for (int n = 0; n < pages; ++n) {
             if (!firstPage)
                   printer.newPage();
             firstPage = false;
+
+            if(usePageColor) {
```
Comment (Jojo-Schmitz) [inline mscore/file.cpp:2152]: space missing

File: mscore/preferences.cpp (hunk 2 above)
Comment (Jojo-Schmitz) [reply mscore/preferences.cpp:160]: Maybe not, but why then are the vtests show background color artifacts?

Comment (SKefalidis) [reply mscore/preferences.cpp:160]: vtests are happy now

File: mscore/file.cpp (hunk 3 above)
Comment (SKefalidis) [reply mscore/file.cpp]: fixed

File: mscore/file.cpp (hunk 4 above)
Comment (SKefalidis) [reply mscore/file.cpp]: fixed

**Comment by Philzen** [comment]
Awesome that you're taking this up :+1: 

> The page color still is a preference. I could change that and make it part of each Page object (as proposed in the issue).

If it's not too much of a kerfuffle, i would go for that. It seems only logical for preferences to be concerned only with application wide-settings (such as the background color of the viewport), while page-settings ("paper color") would be a score property (as  `<pageWidth>`, `<pageHeight>`, etc. are also properties of the generated score XML).

Having said that, it may still be useful to leave it in the preferences as "*Default* Paper color" that will be used when no individual paper color was defined for the score.

I cannot judge the gravity of such a change - it could as well be done as a second iteration. Anyway i'd like to hear what the veterans have to say.

File: mscore/file.cpp [hunk 5]
```diff
@@ -2143,10 +2143,18 @@ bool MuseScore::savePdf(Score* cs_, QPrinter& printer)
       const QList<Page*> pl = cs_->pages();
       int pages = pl.size();
       bool firstPage = true;
+      bool usePageColor = preferences.getBool(PREF_UI_CANVAS_FG_USECOLOR_IN_EXPORT);
       for (int n = 0; n < pages; ++n) {
             if (!firstPage)
                   printer.newPage();
             firstPage = false;
+
+            if (usePageColor) {
```
Comment (Philzen) [inline mscore/file.cpp:2152]: I sense a code smell here. The code itself is fine, but the fact that it's is copied and pasted two more times is not (as it contradicts the [DRY](https://de.wikipedia.org/wiki/Don%E2%80%99t_repeat_yourself) principle). 

My approach would be to move your code into a private subroutine (i.e. something like `applyPageColor(Printer, pageNum)`) and just call that.

But i can see why that that happened: At first glance it looks to me as if `savePdf(Score* cs_, QPrinter& printer)` and `savePdf(QList<Score*> cs_, const QString& saveName)` already have a serious amount of code duplication (if i'm not misunderstanding something here and they do something completely), i'd advise to refactor this at some point in the near future to reduce technical debt and make the code more stable and easier to maintain in the future.

**Comment by Philzen** [comment]
One more thought: The Preferences call it "*Paper*", the variable used in the pull request under discussion refers to it as "*Page*" (as we're also discussing whether to move the UI component to "*Page*(s) Settings", while the Qt IDs refer to it as "Foreground")  ;)

"Paper" is a nice metaphore, but maybe not the ideal one as we're mostly talking about "virtual" paper at best here...

Suggestion: rename the UI setting to "Page Color" to reduce ambiguity?

**Comment by Philzen** [comment]
This new extra checkbox is unnecessary imho:
new export checkbox

Also it deals with exporting and thus logically belongs into that respective tab - which basically already has such a checkbox for PNG, which ([as pointed out in #303756](https://musescore.org/en/node/303756#comment-991315)) currently only gives the option to export in white or ([partially broken](https://musescore.org/en/node/303754) transparency.

If we get [#303760](https://musescore.org/en/node/303760) implemented in a followup incrementation, i believe that would solve potential user confusion in their export workflow as it would provide a clearer, less ambiguous way of configuring exports, so right now my suggestion is to have this PR include:

1. page (paper) background color export :heavy_check_mark: 
2. rename "paper" to "page" color in the UI
3. (optionally) move that setting in the page settings dialog, given everybody agrees this is the way to go. I understand this also requires to move it out of the PREFS_UI* context so it becomes a score property, and that it must then also be stored in the individual score file - variant (a).

Other variants of 3. would be to not remove the preference but rename it to either (b) "page display color" or (c) "page default color", where (b) would make it clear that it's just the color it's displayed in, but not the actual page color (which would then be defined in score page settings), or (c) where it would be the default page color for new scores being created.

So maybe that would need some discussion and find an agreement in the design chat first.

**Comment by SKefalidis** [comment]
Added wallpaper export.

**Comment by SKefalidis** [comment]
> That new "Export" checkbox is unnecessary imho:
> new export checkbox
> 
> It may introduce confusion, as it deals with exporting and thus would logically belong onto the "Export" tab - which has a similar checkbox for PNG already, which ([as pointed out in #303756](https://musescore.org/en/node/303756#comment-991315)) currently only gives the option to export in ([partially broken](https://musescore.org/en/node/303754)) transparency or plain white (a limitation that the PR under discussion fixes).
> 
> If we get [#303760](https://musescore.org/en/node/303760) implemented in a followup increment, i believe that would solve current potential user confusion in their export workflow as it would provide a clearer, less ambiguous way of configuring exports ad-hoc, so right now my suggestion is to have this PR just include 1. + 2. of the following list:
> 
>     1. page (paper) background color export heavy_check_mark
> 
>     2. rename "paper" to "page" color in the UI
> 
>     3. (optionally) move that setting into the page settings dialog, given everybody agrees this is the way to go. I understand this also requires to move it out of the PREFS_UI* context so it becomes a score property, and that it must then also be stored in the individual score file. But this is only one of three possible variants i can think of (a).
> 
> 
> Other variants of 3. would be not removing the preference option but renaming it to either (b) "page display color" or (c) "page default color", where (b) would make it clear that it's just the color it's displayed in, but not the actual page color (which would then be defined in score page settings), or (c) where it would be the default page color for new scores being created.
> 
> So maybe that would need some discussion and find an agreement in the design chat first.

I would like to keep this PR about adding the functionality and leave the proposed UI changes (with which I agree) for another PR.

File: mscore/file.cpp [hunk 6]
```diff
@@ -2155,6 +2155,12 @@ bool MuseScore::savePdf(Score* cs_, QPrinter& printer)
                         p.fillRect(pl.at(n)->pageBoundingRect(), preferences.getColor(PREF_UI_CANVAS_FG_COLOR));
                         p.drawRect(pl.at(n)->pageBoundingRect());
                         }
+                  else {
+                        QPixmap* pm = new QPixmap(preferences.getString(PREF_UI_CANVAS_FG_WALLPAPER));
+                        p.drawTiledPixmap(pl.at(n)->pageBoundingRect(),
+                                          *pm,
+                                          pl.at(n)->pageBoundingRect().topLeft() - QPoint(lrint(mscore->currentScoreView()->matrix().dx()), lrint(mscore->currentScoreView()->matrix().dy())));
```
Comment (Jojo-Schmitz) [inline mscore/file.cpp:2162]: Add one more line break there before the 4th arg.

File: mscore/file.cpp [hunk 7]
```diff
@@ -2243,6 +2249,12 @@ bool MuseScore::savePdf(QList<Score*> cs_, const QString& saveName)
                               p.fillRect(pl.at(n)->pageBoundingRect(), preferences.getColor(PREF_UI_CANVAS_FG_COLOR));
                               p.drawRect(pl.at(n)->pageBoundingRect());
                               }
+                        else {
+                              QPixmap* pm = new QPixmap(preferences.getString(PREF_UI_CANVAS_FG_WALLPAPER));
+                              p.drawTiledPixmap(pl.at(n)->pageBoundingRect(),
+                                                *pm,
+                                                pl.at(n)->pageBoundingRect().topLeft() - QPoint(lrint(mscore->currentScoreView()->matrix().dx()), lrint(mscore->currentScoreView()->matrix().dy())));
```
Comment (Jojo-Schmitz) [inline mscore/file.cpp:2256]: dito

File: mscore/file.cpp [hunk 8]
```diff
@@ -2740,6 +2752,12 @@ bool MuseScore::savePng(Score* score, QIODevice* device, int pageNumber, bool dr
             p.fillRect(pl.at(pageNumber)->pageBoundingRect(), preferences.getColor(PREF_UI_CANVAS_FG_COLOR));
             p.drawRect(pl.at(pageNumber)->pageBoundingRect());
             }
+      else if (useCustomPage) {
+            QPixmap* pm = new QPixmap(preferences.getString(PREF_UI_CANVAS_FG_WALLPAPER));
+            p.drawTiledPixmap(pl.at(pageNumber)->pageBoundingRect(),
+                              *pm,
+                              pl.at(pageNumber)->pageBoundingRect().topLeft() - QPoint(lrint(mscore->currentScoreView()->matrix().dx()), lrint(mscore->currentScoreView()->matrix().dy())));
```
Comment (Jojo-Schmitz) [inline mscore/file.cpp:2759]: dito

File: mscore/file.cpp [hunk 9]
```diff
@@ -2963,8 +2981,15 @@ bool MuseScore::saveSvg(Score* score, QIODevice* device, int pageNumber, bool dr
...
             if (useCustomPage) {
-                  if(preferences.getBool(PREF_UI_CANVAS_FG_USECOLOR))
+                  if(preferences.getBool(PREF_UI_CANVAS_FG_USECOLOR)) {
                         p.fillRect(r, preferences.getColor(PREF_UI_CANVAS_FG_COLOR));
+                        }
+                  else {
+                        QPixmap* pm = new QPixmap(preferences.getString(PREF_UI_CANVAS_FG_WALLPAPER));
+                        p.drawTiledPixmap(pl.at(pageNumber)->pageBoundingRect(),
+                                          *pm,
+                                          pl.at(pageNumber)->pageBoundingRect().topLeft() - QPoint(lrint(mscore->currentScoreView()->matrix().dx()), lrint(mscore->currentScoreView()->matrix().dy())));
```
Comment (Jojo-Schmitz) [inline mscore/file.cpp:2991]: dito

File: mscore/file.cpp (hunk 6 above)
Comment (SKefalidis) [reply mscore/file.cpp]: There is no 4th argument, but I see that it is quite convoluted. It should be clearer now.

File: mscore/file.cpp (hunk 5 above)
Comment (SKefalidis) [reply mscore/file.cpp]: Taken care of :-)

File: mscore/file.cpp (hunk 6 above)
Comment (SKefalidis) [reply mscore/file.cpp]: Hm... the third argument might be both useless and an obstacle in the future, I'll remove it.

**Comment by SKefalidis** [comment]
Cleanup/refactoring.

**Comment by igorkorsukov** [comment]
Unfortunately, now not actual

**Pull request closed without merging**